File: code/go/day22/day22.go
package day22

import "../utils/linkedlist"

// 计算环的长度
func LengthOfLoop(ll *linkedlist.List) int {
	head := ll.GetHead()
	p1 := head.GetNext() // 移动慢的指针，一次一步
	p2 := p1.GetNext()   // 移动快的指针，一次两步
	length := 0          // 环的长度
	hasLoop := false

	for p1 != nil && p2 != nil {
		//p2追上p1, 说明存在环
		if p1 == p2 {
			hasLoop = true
			break
		}

		p1 = p1.GetNext()
		p2 = p2.GetNext()
		if p2 != nil {
			p2 = p2.GetNext()
		}
	}

	// 存在环的话，此时p2不动，p1在以1的步速移动。
	// 统计移动次数，直到与p2相遇,移动的次数即为环大小。
	if hasLoop {
		length = 1
		p1 = p1.GetNext()
		for p1 != p2 {
			length++
			p1 = p1.GetNext()
		}
	}
	return length
}

// 寻找环的入口节点
func EntryNodeOfLoop(ll *linkedlist.List) *linkedlist.Node {
	length := LengthOfLoop(ll)

	// 不存在环
	if length == 0 {
		return nil
	}

	// 申明两个指针，之间差了环的长度个节点
	p1 := ll.GetHead()
	p2 := p1
	for i := 0; i < length; i++ {
		p2 = p2.GetNext()
	}

	for p1 != p2 {
		p1 = p1.GetNext()
		p2 = p2.GetNext()
	}

	return p1

}

File: code/python/day22/day22.py
import unittest

from utils.linked_list import LinkedList, Node


def get_length_of_loop_link(linklist):
    # setp1 是否有环， p1走一步， p2走两步，若相遇则有环
    p1 = p2 = linklist.head
    while p2 is not None:
        p2 = p2.get_next()
        if p2 is None:
            return -1

        p2 = p2.get_next()
        if p2 is None:
            return -1
        p1 = p1.get_next()

        if p1 == p2:
            has_ring = True
            break

    # setp2 求环长度length
    length = 1
    while p2.get_next() != p1:
        p2 = p2.get_next()
        length += 1

    # setp3 p1走l+1步，p2 走1步， 两者相遇即为入口
    p1 = p2 = linklist.head
    for i in range(length):
        p1 = p1.get_next()

    while p1 != p2:
        p1 = p1.get_next()
        p2 = p2.get_next()

    return p1.get_data()

class TestFindKthToTail(unittest.TestCase):
    def test_1(self):
        ll = LinkedList()
        node1 = Node(1)
        node2 = Node(2)
        node3 = Node(3)
        node4 = Node(4)
        node5 = Node(5)
        ll.append_node(node1)
        ll.append_node(node2)
        ll.append_node(node3)
        ll.append_node(node4)
        ll.append_node(node5)
        ll.append_node(node3)
        self.assertEqual(get_length_of_loop_link(ll), 3)

    def test_2(self):
        ll = LinkedList()
        ll.append(1)
        self.assertEqual(get_length_of_loop_link(ll), -1)


if __name__ == '__main__':
    unittest.main()
